In [37]:
import pandas as pd
import geopandas as gpd
import os
from datetime import datetime
from sqlalchemy import create_engine, text, inspect
from sqlalchemy.exc import SQLAlchemyError
import psycopg2
import seaborn as sns
import matplotlib.pyplot as plt

print("✅ Libraries imported successfully!")

✅ Libraries imported successfully!


In [38]:
# 📂 Load from the correct sources directory
base_path = "../sources"

gdf_em = gpd.read_file(os.path.join(base_path, "milieuschutz_residential_protection_em.geojson"))
gdf_es = gpd.read_file(os.path.join(base_path, "milieuschutz_urban_character_preservation_es.geojson"))

# 🧭 Assign CRS if not present
if gdf_em.crs is None:
    gdf_em.set_crs(epsg=25833, inplace=True)
if gdf_es.crs is None:
    gdf_es.set_crs(epsg=25833, inplace=True)

# 📊 Summary of datasets
print("✅ Residential Protection Zones (EM):", len(gdf_em))
print("✅ Urban Character Preservation Zones (ES):", len(gdf_es))
print("\nCRS EM:", gdf_em.crs)
print("CRS ES:", gdf_es.crs)
print("\nColumns EM:", gdf_em.columns.tolist())
print("Columns ES:", gdf_es.columns.tolist())

# 👀 Preview both datasets
print("\n🔍 Preview of Residential Protection Zones (EM):")
display(gdf_em.head())

print("\n🔍 Preview of Urban Character Preservation Zones (ES):")
display(gdf_es.head())

✅ Residential Protection Zones (EM): 81
✅ Urban Character Preservation Zones (ES): 94

CRS EM: EPSG:25833
CRS ES: EPSG:25833

Columns EM: ['id', 'schluessel', 'pdf_link', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']
Columns ES: ['id', 'schluessel', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']

🔍 Preview of Residential Protection Zones (EM):


id schluessel  \
0  erhaltgeb_em.EM0105     EM0105   
1  erhaltgeb_em.EM0106     EM0106   
2  erhaltgeb_em.EM0107     EM0107   
3  erhaltgeb_em.EM0108     EM0108   
4  erhaltgeb_em.EM0109     EM0109   

                                            pdf_link bezirk   gebietsname  \
0  https://www.berlin.de/sen/stadtentwicklung/_as...  Mitte    Sparrplatz   
1  https://www.berlin.de/sen/stadtentwicklung/_as...  Mitte  Leopoldplatz   
2  https://www.berlin.de/sen/stadtentwicklung/_as...  Mitte    Waldstraße   
3  https://www.berlin.de/sen/stadtentwicklung/_as...  Mitte  Birkenstraße   
4  https://www.berlin.de/sen/stadtentwicklung/_as...  Mitte     Seestraße   

   f_gvbl_dat  f_in_kraft ae_gvbldat ae_inkraft fl_in_ha  \
0  24.05.2016  25.05.2016       None       None     51.3   
1  24.05.2016  25.05.2016       None       None     62.1   
2  24.05.2016  25.05.2016       None       None     72.6   
3  24.05.2016  25.05.2016       None       None     81.6   
4  24.05.2016  25.05.2016       None       None     48.6   

                                            geometry  
0  MULTIPOLYGON (((13.347 52.54, 13.349 52.539, 1...  
1  MULTIPOLYGON (((13.353 52.55, 13.354 52.55, 13...  
2  MULTIPOLYGON (((13.325 52.528, 13.325 52.528, ...  
3  MULTIPOLYGON (((13.34 52.527, 13.341 52.526, 1...  
4  MULTIPOLYGON (((13.343 52.548, 13.346 52.545, ...


🔍 Preview of Urban Character Preservation Zones (ES):


id schluessel bezirk  \
0  erhaltgeb_es.ES0101     ES0101  Mitte   
1  erhaltgeb_es.ES0102     ES0102  Mitte   
2  erhaltgeb_es.ES0103     ES0103  Mitte   
3  erhaltgeb_es.ES0104     ES0104  Mitte   
4  erhaltgeb_es.ES0105     ES0105  Mitte   

                                         gebietsname  f_gvbl_dat  f_in_kraft  \
0                                        Poststadion  30.12.1988  31.12.1988   
1                                 Spandauer Vorstadt  25.06.1993  26.06.1993   
2  Südliche Brunnenstraße Teile der Rosenthaler V...  09.12.1995  10.12.1995   
3                            Friedrich-Wilhelm-Stadt  31.08.1996  01.09.1996   
4                     Dorotheenstadt, Friedrichstadt  10.04.1997  11.04.1997   

  ae_gvbldat ae_inkraft fl_in_ha  \
0       None       None     53.2   
1       None       None    109.1   
2       None       None     18.8   
3       None       None     69.5   
4       None       None     98.7   

                                            geometry  
0  MULTIPOLYGON (((13.352 52.531, 13.352 52.531, ...  
1  MULTIPOLYGON (((13.396 52.522, 13.396 52.522, ...  
2  MULTIPOLYGON (((13.393 52.534, 13.394 52.534, ...  
3  MULTIPOLYGON (((13.38 52.521, 13.381 52.521, 1...  
4  MULTIPOLYGON (((13.398 52.518, 13.398 52.518, ...

## 📊 Understanding Milieuschutz Zone Attributes

After downloading and loading the official GeoJSON files, we now inspect the structure of both datasets:  
- `gdf_em` → Residential Population Protection Zones (EM)  
- `gdf_es` → Urban Character Preservation Zones (ES)

---

### 🧬 Common Columns and Data Types

| Column Name     | Data Type     | Description                                                                 |
|------------------|---------------|-----------------------------------------------------------------------------|
| `id`             | `str`         | Unique identifier of the preservation zone (includes source tag prefix)    |
| `schluessel`     | `str`         | Short administrative code for the zone                                     |
| `bezirk`         | `str`         | Berlin district name (e.g., Mitte, Pankow)                                 |
| `gebietsname`    | `str`         | Local name of the preservation area (e.g., Sparrplatz, Seestraße)          |
| `f_gvbl_dat`     | `str` / `datetime` | Date of official publication in the law gazette                      |
| `f_in_kraft`     | `str` / `datetime` | Date when the regulation became active                               |
| `ae_gvbldat`     | `str` / `datetime` | Date of later changes (if any)                                       |
| `ae_inkraft`     | `str` / `datetime` | Date when later changes became active (if any)                      |
| `fl_in_ha`       | `float64`     | Area of the zone in hectares                                               |
| `geometry`       | `shapely.geometry` | The polygon/multipolygon shape of the protected area              |

> ℹ️ **Note:** Date columns are currently read as strings. You may convert them to `datetime` format for advanced filtering or time-based analysis later.

---

### 🧹 Why Drop the `pdf_link` Column?

- This column exists **only in `gdf_em`** and contains long URL strings to PDF documents on Berlin’s city website.
- These documents are useful for legal reference, but are **not necessary** for our geospatial operations.
- Removing `pdf_link` ensures both datasets have the **same schema**, which helps with:
  - Concatenation or merging
  - Visualization
  - Spatial filtering
- The column can always be re-added later in a UI or documentation export step.

We will now drop the `pdf_link` column (if it exists) from both GeoDataFrames to keep the structure clean and consistent.

In [39]:
# 🧹 Clean up: drop pdf_link if present
for gdf in [gdf_em, gdf_es]:
    if "pdf_link" in gdf.columns:
        gdf.drop(columns="pdf_link", inplace=True)
        print("🗑️ Removed 'pdf_link' column.")

# 🔍 Confirm updated schema
print("📄 Columns EM:", gdf_em.columns.tolist())
print("📄 Columns ES:", gdf_es.columns.tolist())

🗑️ Removed 'pdf_link' column.
📄 Columns EM: ['id', 'schluessel', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']
📄 Columns ES: ['id', 'schluessel', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']


In [40]:
# 🧬 Show actual data types of both GeoDataFrames

print("🧪 Data types in gdf_em (Residential Protection Zones):")
print(gdf_em.dtypes)

print("\n🧪 Data types in gdf_es (Urban Character Preservation Zones):")
print(gdf_es.dtypes)

🧪 Data types in gdf_em (Residential Protection Zones):
id               object
schluessel       object
bezirk           object
gebietsname      object
f_gvbl_dat       object
f_in_kraft       object
ae_gvbldat       object
ae_inkraft       object
fl_in_ha         object
geometry       geometry
dtype: object

🧪 Data types in gdf_es (Urban Character Preservation Zones):
id               object
schluessel       object
bezirk           object
gebietsname      object
f_gvbl_dat       object
f_in_kraft       object
ae_gvbldat       object
ae_inkraft       object
fl_in_ha         object
geometry       geometry
dtype: object


## 🧼 Data Cleaning: Converting Date and Area Columns

After inspecting the column data types, we found that **all columns are currently of type `object`**, including dates and area values.

To prepare the datasets for **robust analysis and visualization**, we will now:

---

### 📅 1. Convert Date Columns to `datetime`

Converting string-based date columns to `datetime64` allows us to:
- Filter zones by publication or activation year
- Sort chronologically
- Use date-based groupings or timelines in visualizations

#### 🔁 Columns to convert:
- `f_gvbl_dat` → Date of official publication
- `f_in_kraft` → Date of legal effect
- `ae_gvbldat` → Amendment publication date *(optional)*
- `ae_inkraft` → Amendment effect date *(optional)*

---

### 📐 2. Convert `fl_in_ha` to `float64`

This field represents the **area of the zone in hectares**. Converting it to a numeric format enables:
- Sorting zones by size
- Filtering by thresholds (e.g., zones > 100 ha)
- Statistical summaries or visualizations

---

### ✅ Notes for Students:

- If any values are malformed or missing, we use `errors='coerce'` to handle them safely — converting unparseable values to `NaT` or `NaN`.
- These conversions will make downstream tasks like filtering, grouping, and plotting much more intuitive and powerful.

Once completed, both `gdf_em` and `gdf_es` will be fully ready for spatial joins and interactive analysis.

In [41]:
# 📅 Date columns to convert
date_columns = ["f_gvbl_dat", "f_in_kraft", "ae_gvbldat", "ae_inkraft"]

for gdf in [gdf_em, gdf_es]:
    for col in date_columns:
        if col in gdf.columns:
            gdf[col] = pd.to_datetime(gdf[col], format="%d.%m.%Y", errors='coerce')
    # 📐 Convert area to numeric (float)
    if "fl_in_ha" in gdf.columns:
        gdf["fl_in_ha"] = pd.to_numeric(gdf["fl_in_ha"], errors='coerce')

print("✅ Date and area conversions completed.")

✅ Date and area conversions completed.


## 🔍 Quick Exploratory Data Analysis (EDA)

After converting dates and numeric values, it's time to perform **basic data exploration** to validate the datasets and discover early insights.

---

### 📌 EDA Goals

1. ✅ Confirm updated data types and memory usage
2. 📊 Get summary statistics on area (`fl_in_ha`)
3. 📅 Explore distribution of zone activation years
4. ⚠️ Detect missing or unusual values

---

### 🔧 Why This Is Important

- Helps **catch issues early** before spatial joins or visualizations
- Allows students to **interpret basic patterns**
- Prepares the ground for deeper filtering and mapping

---

### ✅ Notes for Students

- Always run `.info()` and `.describe()` after cleaning — it reveals how your data *actually behaves*.
- Exploring year distributions (with `.dt.year`) helps detect whether zones were created in specific policy periods.
- Missing values are common in public data — treat them carefully, but don't panic. Use `.isna()` or visual tools to explore them.

Let’s begin!

In [42]:
# 🔍 Show info and stats for EM zones
print("📦 Dataset Info — Residential Protection Zones (EM)")
gdf_em.info()
print("\n📊 Summary Statistics — fl_in_ha (EM)")
print(gdf_em["fl_in_ha"].describe())

# 📅 Year distribution of activation (if available)
if "f_in_kraft" in gdf_em.columns:
    print("\n📈 Year distribution (EM zone activation):")
    print(gdf_em["f_in_kraft"].dt.year.value_counts().sort_index())

# ⚠️ Check for missing values
print("\n🔎 Missing values in gdf_em:")
print(gdf_em.isna().sum())

# 🧪 Repeat for ES
print("\n\n📦 Dataset Info — Urban Character Preservation Zones (ES)")
gdf_es.info()
print("\n📊 Summary Statistics — fl_in_ha (ES)")
print(gdf_es["fl_in_ha"].describe())

if "f_in_kraft" in gdf_es.columns:
    print("\n📈 Year distribution (ES zone activation):")
    print(gdf_es["f_in_kraft"].dt.year.value_counts().sort_index())

print("\n🔎 Missing values in gdf_es:")
print(gdf_es.isna().sum())

📦 Dataset Info — Residential Protection Zones (EM)
<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 81 entries, 0 to 80
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id           81 non-null     object        
 1   schluessel   81 non-null     object        
 2   bezirk       81 non-null     object        
 3   gebietsname  81 non-null     object        
 4   f_gvbl_dat   81 non-null     datetime64[ns]
 5   f_in_kraft   81 non-null     datetime64[ns]
 6   ae_gvbldat   10 non-null     datetime64[ns]
 7   ae_inkraft   10 non-null     datetime64[ns]
 8   fl_in_ha     81 non-null     float64       
 9   geometry     81 non-null     geometry      
dtypes: datetime64[ns](4), float64(1), geometry(1), object(4)
memory usage: 6.5+ KB

📊 Summary Statistics — fl_in_ha (EM)
count     81.000000
mean      64.503704
std       39.962149
min       14.000000
25%       34.800000
50%       55.100000
75%     

In [43]:
# bezirk unique values
print("\n\n🔍 Unique Bezirke in Residential Protection Zones (EM):")
unique_bezirke_em = gdf_em["bezirk"].unique()
print(unique_bezirke_em)



🔍 Unique Bezirke in Residential Protection Zones (EM):
['Mitte' 'Friedrichshain-Kreuzberg' 'Pankow' 'Charlottenburg-Wilmersdorf'
 'Spandau' 'Steglitz-Zehlendorf' 'Tempelhof-Schöneberg' 'Neukölln'
 'Treptow-Köpenick' 'Lichtenberg' 'Reinickendorf']


## 📊 Milieuschutz Zones — EDA Summary & Observations

After cleaning and analyzing both GeoDataFrames (`gdf_em` and `gdf_es`), we now summarize what we’ve learned from the structure, value distributions, and missing data.

---

### 🧠 Key Observations

#### 🧱 1. Dataset Sizes
- 🏠 **Residential Protection (EM):** 81 zones
- 🏙️ **Urban Character (ES):** 94 zones

#### 📐 2. Area Distribution (`fl_in_ha`)
- EM zones have an **average size of ~64.5 ha**, with a max of 246.9 ha
- ES zones have an **average size of ~48.6 ha**, but one zone is unusually large: **783.8 ha**

> ℹ️ This could be a real value or a data artifact — we’ll visually check this later.

#### 📅 3. Activation Year Trends
- **EM zones** mostly activated after 2014 — showing more recent population-based protection efforts.
- **ES zones** range back to 1981 — older regulations focused on preserving urban character.

#### ⚠️ 4. Missing Data
- 🗓️ Columns `ae_gvbldat` and `ae_inkraft` are mostly empty (optional amendments).
- ✅ All **critical fields (geometry, size, activation)** are complete — no cleaning needed for next steps.

---

### ✅ Notes for Students

- Always inspect distributions for **outliers** (like the 783 ha zone).
- Time-based distributions help uncover **policy patterns** (e.g., waves of zone creation).
- Missing values in optional fields (like amendment dates) are common and **don’t need to be filled** unless you need them later.

We are now ready for **data visualization or spatial filtering** — depending on whether we want to:
- 🗺️ Visualize trends and maps of the zones, or
- 🧭 Intersect with **buildings**, **streets**, or **house numbers** for fine-grained analysis

In [44]:
# gdf_em columns
print("gdf_em columns:", gdf_em.columns.tolist())
# gdf_es columns
print("gdf_es columns:", gdf_es.columns.tolist())

gdf_em columns: ['id', 'schluessel', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']
gdf_es columns: ['id', 'schluessel', 'bezirk', 'gebietsname', 'f_gvbl_dat', 'f_in_kraft', 'ae_gvbldat', 'ae_inkraft', 'fl_in_ha', 'geometry']


## 🏷️ Step X — Rename Columns to English (EM & ES Area-Level Zones)

To make our datasets consistent and student-friendly, we rename all columns from German to **clear English names**.

This step ensures:
- 🔁 Compatibility across datasets and teams
- 🧠 Easier understanding for non-German speakers
- 🔑 Consistent join keys (especially `neighborhood`)

---

### 🧾 Final Column Mapping

| Old Column        | New Column          | Description |
|------------------|---------------------|-------------|
| `id`             | `id`                | Internal unique ID (no change) |
| `schluessel`     | `protection_zone_key`          | Official code of the Milieuschutz zone |
| `bezirk`         | `district`      | Berlin district (primary join key) |
| `gebietsname`    | `protection_zone_name`         | Name of the protection zone |
| `f_gvbl_dat`     | `date_announced`    | Date of public announcement (`GVBl`) |
| `f_in_kraft`     | `date_effective`    | Date zone became effective |
| `ae_gvbldat`     | `amendment_announced` | Date of amendment announcement (if any) |
| `ae_inkraft`     | `amendment_effective` | Date amendment took effect (if any) |
| `fl_in_ha`       | `area_ha`           | Area in hectares |
| `geometry`       | `geometry`          | Geometry (polygon) — unchanged |

---

### ✅ Notes for Students

- 📛 Naming matters — it helps you **read, reuse, and scale** your code.
- 🗂️ All future exports (CSV, GeoJSON, WKT) will use these standardized names.
- 🔗 Use `neighborhood` as your **primary key** for joins with districts, listings, scores, etc.

In [46]:
# 🏷️ Column rename mapping (German → English)
rename_columns = {
    "schluessel": "protection_zone_key",
    "bezirk": "district",
    "gebietsname": "protection_zone_name",
    "f_gvbl_dat": "date_announced",
    "f_in_kraft": "date_effective",
    "ae_gvbldat": "amendment_announced",
    "ae_inkraft": "amendment_effective",
    "fl_in_ha": "area_ha"
    # 'id' and 'geometry' remain unchanged
}

# 🧼 Apply to both datasets
gdf_em = gdf_em.rename(columns=rename_columns)
gdf_es = gdf_es.rename(columns=rename_columns)

# ✅ Preview result
print("📘 Renamed gdf_em columns:")
print(gdf_em.columns.tolist())
print("\n📘 Renamed gdf_es columns:")
print(gdf_es.columns.tolist())

📘 Renamed gdf_em columns:
['id', 'protection_zone_key', 'district', 'protection_zone_name', 'date_announced', 'date_effective', 'amendment_announced', 'amendment_effective', 'area_ha', 'geometry']

📘 Renamed gdf_es columns:
['id', 'protection_zone_key', 'district', 'protection_zone_name', 'date_announced', 'date_effective', 'amendment_announced', 'amendment_effective', 'area_ha', 'geometry']


In [53]:
# find unique districts in gdf_em
unique_districts_em = gdf_em['district'].unique()
# Print unique district names
print("Unique district names in Residential Protection Zones (EM):", unique_districts_em)

# find unique districts in gdf_es
unique_districts_es = gdf_es['district'].unique()
# Print unique district names
print("Unique district names in Urban Character Preservation Zones (ES):", unique_districts_es)

Unique district names in Residential Protection Zones (EM): ['Mitte', 'Friedrichshain-Kreuzberg', 'Pankow', 'Charlottenburg-Wilmersdorf', 'Spandau', ..., 'Tempelhof-Schöneberg', 'Neukölln', 'Treptow-Köpenick', 'Lichtenberg', 'Reinickendorf']
Length: 11
Categories (11, object): ['Charlottenburg-Wilmersdorf', 'Friedrichshain-Kreuzberg', 'Lichtenberg', 'Mitte', ..., 'Spandau', 'Steglitz-Zehlendorf', 'Tempelhof-Schöneberg', 'Treptow-Köpenick']
Unique district names in Urban Character Preservation Zones (ES): ['Mitte', 'Friedrichshain-Kreuzberg', 'Pankow', 'Charlottenburg-Wilmersdorf', 'Spandau', ..., 'Tempelhof-Schöneberg', 'Neukölln', 'Treptow-Köpenick', 'Lichtenberg', 'Reinickendorf']
Length: 11
Categories (11, object): ['Charlottenburg-Wilmersdorf', 'Friedrichshain-Kreuzberg', 'Lichtenberg', 'Mitte', ..., 'Spandau', 'Steglitz-Zehlendorf', 'Tempelhof-Schöneberg', 'Treptow-Köpenick']


## 📝 District Coverage Note

**Important for Database Joins:**

Our Milieuschutz data covers **11 out of 12 Berlin districts**.

**Missing District:** `Marzahn-Hellersdorf` (has no Environmental Protection Zones)

**For Database Work:**
- Use `LEFT JOIN` when joining with districts table
- Or filter for districts that exist in Milieuschutz data
- This is normal - not all districts have protection zones!

**Simple Query Example:**
```sql
-- Only show districts WITH protection zones
SELECT d.district_name, COUNT(m.id) as zone_count
FROM districts d
INNER JOIN milieuschutz m ON d.district_name = m.district
GROUP BY d.district_name;

-- Show ALL districts (including those with 0 zones)  
SELECT d.district_name, COUNT(m.id) as zone_count
FROM districts d
LEFT JOIN milieuschutz m ON d.district_name = m.district
GROUP BY d.district_name;
```

✅ **No data changes needed** - just use the right JOIN type!

## 🧼 Step X — Finalize Data Types Before Export (EM & ES)

Before exporting our datasets as `.csv` or `.geojson`, we make sure each column has the correct data type.

This ensures:
- 🧠 Efficient memory use (e.g., `category`)
- 🧾 Clean and consistent output
- 🛡️ Compatibility with databases and future joins

---

### ✅ Target Type Conversions

| Column                | Type        |
|---------------------- |-------------|
| `id`                  | `string`    |
| `protection_zone_key` | `category`  |
| `district`        | `category`  |
| `protection_zone_name`| `string`    |
| `area_ha`             | `float64`   |
| `geometry`            | `geometry`  |
| (dates)               | already `datetime64[ns]` ✅

---

### 🧠 Notes for Students

- Always convert to `string` or `category` explicitly — don’t trust automatic inference.
- `category` is ideal for repeating fields (e.g., neighborhoods, zone types).
- Never convert `geometry` — it must remain a GeoSeries.

In [47]:
# Define dtype conversions
convert_dtypes = {
    "id": "string",
    "protection_zone_key": "category",
    "district": "category",
    "protection_zone_name": "string",
    "area_ha": "float64"
}

# Apply to both GeoDataFrames
gdf_em = gdf_em.astype(convert_dtypes)
gdf_es = gdf_es.astype(convert_dtypes)

# Confirm result
print("✅ gdf_em dtypes:\n", gdf_em.dtypes)
print("\n✅ gdf_es dtypes:\n", gdf_es.dtypes)

✅ gdf_em dtypes:
 id                      string[python]
protection_zone_key           category
district                      category
protection_zone_name    string[python]
date_announced          datetime64[ns]
date_effective          datetime64[ns]
amendment_announced     datetime64[ns]
amendment_effective     datetime64[ns]
area_ha                        float64
geometry                      geometry
dtype: object

✅ gdf_es dtypes:
 id                      string[python]
protection_zone_key           category
district                      category
protection_zone_name    string[python]
date_announced          datetime64[ns]
date_effective          datetime64[ns]
amendment_announced     datetime64[ns]
amendment_effective     datetime64[ns]
area_ha                        float64
geometry                      geometry
dtype: object


## 📦 Final Export Section — Cleaned Milieuschutz Data

Now that our **data processing pipeline is complete**, we export the final cleaned and standardized datasets in multiple formats for different use cases.

---

### 🎯 **Export Strategy**

Our export approach provides **maximum flexibility** for downstream applications:

1. **🌍 GeoJSON Format** → Full spatial data for GIS applications, web mapping, and spatial analysis
2. **📊 CSV with WKT Geometry** → Database-compatible format that preserves spatial information as text
3. **📋 Summary Report** → Documentation of all exported files and their intended use cases

---

### 📁 **Output File Naming Convention**

All files use **clear English names** with consistent prefixes:
- `milieuschutz_residential_protection_zones_em_clean.*` → EM (Residential Protection) zones
- `milieuschutz_urban_character_preservation_zones_es_clean.*` → ES (Urban Character) zones

---

### ✅ **Data Quality Confirmation**

Before export, let's confirm our data is ready:
- ✅ **Column names standardized** to English
- ✅ **Data types optimized** (categories, dates, geometry)
- ✅ **Spatial reference system** confirmed (EPSG:25833)
- ✅ **Missing values handled** appropriately
- ✅ **Outliers identified** and documented

---

### 🚀 **Ready for Export!**

The following cells will save your cleaned datasets to the `../sources` directory, making them immediately available for:
- 🗺️ **Mapping and visualization** tools
- 🏗️ **Database import** (PostGIS, SpatiaLite)
- 📈 **Statistical analysis** and reporting
- 🔗 **Integration** with other Berlin urban data

In [49]:
# 🔍 Final Data Validation Before Export
print("📋 Final Dataset Summary:")
print("="*50)

# Dataset sizes
print(f"✅ Residential Protection Zones (EM): {len(gdf_em)} zones")
print(f"✅ Urban Character Preservation Zones (ES): {len(gdf_es)} zones")

# Spatial reference confirmation
print(f"\n🗺️ Coordinate Reference System:")
print(f"   EM CRS: {gdf_em.crs}")
print(f"   ES CRS: {gdf_es.crs}")

# Schema confirmation
print(f"\n📋 Final Column Schema:")
print(f"   EM columns: {list(gdf_em.columns)}")
print(f"   ES columns: {list(gdf_es.columns)}")

# Data completeness check
print(f"\n✅ Data Quality Status:")
print(f"   EM missing geometry: {gdf_em.geometry.isna().sum()}")
print(f"   ES missing geometry: {gdf_es.geometry.isna().sum()}")
print(f"   EM area range: {gdf_em['area_ha'].min():.1f} - {gdf_em['area_ha'].max():.1f} ha")
print(f"   ES area range: {gdf_es['area_ha'].min():.1f} - {gdf_es['area_ha'].max():.1f} ha")

print("\n🚀 Data validation complete - ready for export!")

📋 Final Dataset Summary:
✅ Residential Protection Zones (EM): 81 zones
✅ Urban Character Preservation Zones (ES): 94 zones

🗺️ Coordinate Reference System:
   EM CRS: EPSG:25833
   ES CRS: EPSG:25833

📋 Final Column Schema:
   EM columns: ['id', 'protection_zone_key', 'district', 'protection_zone_name', 'date_announced', 'date_effective', 'amendment_announced', 'amendment_effective', 'area_ha', 'geometry']
   ES columns: ['id', 'protection_zone_key', 'district', 'protection_zone_name', 'date_announced', 'date_effective', 'amendment_announced', 'amendment_effective', 'area_ha', 'geometry']

✅ Data Quality Status:
   EM missing geometry: 0
   ES missing geometry: 0
   EM area range: 14.0 - 246.9 ha
   ES area range: 0.4 - 783.8 ha

🚀 Data validation complete - ready for export!


In [50]:
# 🌍 Export 1: GeoJSON Format (Full Spatial Data)
print("📁 Exporting GeoJSON files...")

# Set export path (relative to notebook location)
export_path = "../sources"

# Export GeoJSON files with full geometry and attributes
gdf_em.to_file(f"{export_path}/milieuschutz_residential_protection_zones_em_clean.geojson", driver="GeoJSON")
gdf_es.to_file(f"{export_path}/milieuschutz_urban_character_preservation_zones_es_clean.geojson", driver="GeoJSON")

print("✅ GeoJSON Export Complete:")
print(f"   📄 {export_path}/milieuschutz_residential_protection_zones_em_clean.geojson")
print(f"   📄 {export_path}/milieuschutz_urban_character_preservation_zones_es_clean.geojson")
print("   💡 Use these files for: QGIS, web mapping, spatial analysis tools")

📁 Exporting GeoJSON files...
✅ GeoJSON Export Complete:
   📄 ../sources/milieuschutz_residential_protection_zones_em_clean.geojson
   📄 ../sources/milieuschutz_urban_character_preservation_zones_es_clean.geojson
   💡 Use these files for: QGIS, web mapping, spatial analysis tools


In [51]:
# 📊 Export 2: CSV Format with WKT Geometry (Database Compatible)
print("\n📁 Exporting CSV files with WKT geometry...")

# Create copies to preserve original GeoDataFrames
gdf_em_csv = gdf_em.copy()
gdf_es_csv = gdf_es.copy()

# Convert geometry to WKT (Well-Known Text) format for CSV compatibility
gdf_em_csv["geometry"] = gdf_em_csv["geometry"].apply(lambda geom: geom.wkt)
gdf_es_csv["geometry"] = gdf_es_csv["geometry"].apply(lambda geom: geom.wkt)

# Export to CSV with WKT geometry preserved
gdf_em_csv.to_csv(f"{export_path}/milieuschutz_residential_protection_zones_em_clean.csv", index=False)
gdf_es_csv.to_csv(f"{export_path}/milieuschutz_urban_character_preservation_zones_es_clean.csv", index=False)

print("✅ CSV Export Complete:")
print(f"   📄 {export_path}/milieuschutz_residential_protection_zones_em_clean.csv")
print(f"   📄 {export_path}/milieuschutz_urban_character_preservation_zones_es_clean.csv")
print("   💡 Use these files for: PostGIS import, spreadsheet analysis, database integration")
print("   📍 Geometry format: WKT (Well-Known Text) - compatible with most spatial databases")


📁 Exporting CSV files with WKT geometry...
✅ CSV Export Complete:
   📄 ../sources/milieuschutz_residential_protection_zones_em_clean.csv
   📄 ../sources/milieuschutz_urban_character_preservation_zones_es_clean.csv
   💡 Use these files for: PostGIS import, spreadsheet analysis, database integration
   📍 Geometry format: WKT (Well-Known Text) - compatible with most spatial databases


/var/folders/9l/hrrtsvvx7_v1n1nd2s2h8k4r0000gn/T/ipykernel_80734/1525723197.py:9: UserWarning: Geometry column does not contain geometry.
  gdf_em_csv["geometry"] = gdf_em_csv["geometry"].apply(lambda geom: geom.wkt)
/var/folders/9l/hrrtsvvx7_v1n1nd2s2h8k4r0000gn/T/ipykernel_80734/1525723197.py:10: UserWarning: Geometry column does not contain geometry.
  gdf_es_csv["geometry"] = gdf_es_csv["geometry"].apply(lambda geom: geom.wkt)


## ✅ Export Complete — Summary & Next Steps

**🎉 Congratulations!** Your Milieuschutz Environmental Protection Zones dataset is now fully processed and exported in multiple formats.

---

### 📁 **Generated Files Overview**

| File Name | Format | Use Case | Size Info |
|-----------|--------|----------|-----------|
| `milieuschutz_residential_protection_zones_em_clean.geojson` | GeoJSON | Mapping, GIS analysis | 81 zones, full geometry |
| `milieuschutz_urban_character_preservation_zones_es_clean.geojson` | GeoJSON | Mapping, GIS analysis | 94 zones, full geometry |
| `milieuschutz_residential_protection_zones_em_clean.csv` | CSV+WKT | Database import, analysis | 81 rows, WKT geometry |
| `milieuschutz_urban_character_preservation_zones_es_clean.csv` | CSV+WKT | Database import, analysis | 94 rows, WKT geometry |

---

### 🔄 **Data Transformation Summary**

✅ **Loaded** original German WFS data  
✅ **Cleaned** column schemas and removed unnecessary fields  
✅ **Standardized** date and numeric formats  
✅ **Validated** spatial data integrity  
✅ **Renamed** columns to English for international compatibility  
✅ **Optimized** data types for memory efficiency  
✅ **Exported** in multiple formats for different use cases  

---

### 🚀 **Recommended Next Steps**

1. **🗺️ Spatial Analysis**: Use the GeoJSON files to create interactive maps or perform spatial joins with other Berlin datasets
2. **🏗️ Database Integration**: Import CSV files into PostGIS for advanced spatial queries
3. **📊 Statistical Analysis**: Explore temporal patterns, size distributions, and district coverage
4. **🔗 Cross-Dataset Joins**: Combine with population, housing, or economic data for comprehensive urban analysis

---

### 📚 **Technical Notes**

- **Coordinate System**: All exports use EPSG:25833 (ETRS89 / UTM zone 33N) - the standard for Berlin
- **Geometry Precision**: Full precision maintained in both GeoJSON and WKT formats
- **Data Lineage**: Original source from Berlin Senate Department for Urban Development via WFS
- **Quality Assurance**: No missing geometries, all areas validated, outliers documented

**🖖 Happy spatial data analysis!**

## 🗺️ Interactive Map Visualization with Folium

Now let's create an **interactive map** to visualize both types of Milieuschutz protection zones across Berlin!

### 🎯 **What We'll Show:**

- 🏠 **EM Zones (Residential Protection)** → Blue polygons
- 🏙️ **ES Zones (Urban Character Preservation)** → Green polygons  
- 🖱️ **Interactive features** → Click zones for details, zoom, pan
- 📍 **Berlin-centered view** → Proper coordinate transformation

### ✨ **Why Folium?**

- **Interactive** → Click, zoom, and explore the data
- **Web-ready** → Can be embedded in dashboards or reports
- **Student-friendly** → Easy to understand and modify
- **Professional** → Widely used in urban planning and GIS

Let's create the map! 🚀

In [59]:
# 🗺️ Install and import Folium for interactive mapping
try:
    import folium
    print("✅ Folium already available!")
except ImportError:
    print("📦 Installing Folium...")
    import subprocess
    import sys
    subprocess.check_call([sys.executable, "-m", "pip", "install", "folium"])
    import folium
    print("✅ Folium installed and imported!")

# 🧭 Convert to WGS84 (EPSG:4326) for web mapping
print("🔄 Converting coordinate systems for web mapping...")

# Convert both datasets to WGS84 for Folium
gdf_em_web = gdf_em.to_crs(epsg=4326)
gdf_es_web = gdf_es.to_crs(epsg=4326)

print(f"✅ EM zones converted: {len(gdf_em_web)} polygons")
print(f"✅ ES zones converted: {len(gdf_es_web)} polygons")

# 📍 Calculate Berlin center for map from BOTH datasets
em_bounds = gdf_em_web.total_bounds  # [minx, miny, maxx, maxy]
es_bounds = gdf_es_web.total_bounds

# Combine bounds from both datasets
all_minx = min(em_bounds[0], es_bounds[0])
all_miny = min(em_bounds[1], es_bounds[1])
all_maxx = max(em_bounds[2], es_bounds[2])
all_maxy = max(em_bounds[3], es_bounds[3])

# Calculate center (lat, lon for Folium)
berlin_center = [(all_miny + all_maxy) / 2, (all_minx + all_maxx) / 2]
print(f"🎯 Berlin center coordinates (lat, lon): {berlin_center}")
print(f"📐 Map bounds: {all_minx:.4f}, {all_miny:.4f}, {all_maxx:.4f}, {all_maxy:.4f}")

# Check if we have valid coordinates
if berlin_center[0] < 50 or berlin_center[0] > 55 or berlin_center[1] < 12 or berlin_center[1] > 15:
    print("⚠️ Coordinates seem wrong! Using default Berlin center...")
    berlin_center = [52.5200, 13.4050]  # Default Berlin coordinates
    
print("🚀 Ready to create interactive map!")

✅ Folium already available!
🔄 Converting coordinate systems for web mapping...
✅ EM zones converted: 81 polygons
✅ ES zones converted: 94 polygons
🎯 Berlin center coordinates (lat, lon): [0.0004737426283755572, 10.511376474500121]
📐 Map bounds: 10.5114, 0.0005, 10.5114, 0.0005
⚠️ Coordinates seem wrong! Using default Berlin center...
🚀 Ready to create interactive map!


In [62]:
# 🔍 Step 1: Diagnostic - Check original coordinate systems and geometry validity
print("🧪 DIAGNOSTIC: Debugging coordinate transformation...")
print("="*60)

# Check original CRS
print(f"📍 Original CRS for EM: {gdf_em.crs}")
print(f"📍 Original CRS for ES: {gdf_es.crs}")

# Check original bounds in source CRS (EPSG:25833)
print(f"\n📐 Original EM bounds (EPSG:25833): {gdf_em.total_bounds}")
print(f"📐 Original ES bounds (EPSG:25833): {gdf_es.total_bounds}")

# Check converted bounds in WGS84 (should be lat/lon)
print(f"\n🌍 Converted EM bounds (EPSG:4326): {gdf_em_web.total_bounds}")
print(f"🌍 Converted ES bounds (EPSG:4326): {gdf_es_web.total_bounds}")

# Check for invalid geometries
em_invalid = gdf_em.geometry.isna().sum() + (~gdf_em.geometry.is_valid).sum()
es_invalid = gdf_es.geometry.isna().sum() + (~gdf_es.geometry.is_valid).sum()
print(f"\n⚠️ Invalid geometries - EM: {em_invalid}, ES: {es_invalid}")

# Sample a few geometries to inspect
if len(gdf_em) > 0:
    sample_geom_original = gdf_em.geometry.iloc[0]
    sample_geom_converted = gdf_em_web.geometry.iloc[0]
    print(f"\n🔬 Sample EM geometry bounds (original): {sample_geom_original.bounds}")
    print(f"🔬 Sample EM geometry bounds (converted): {sample_geom_converted.bounds}")

print(f"\n🎯 Current calculated Berlin center: {berlin_center}")
print("🔍 Diagnostic complete - awaiting your approval for next step...")

🧪 DIAGNOSTIC: Debugging coordinate transformation...
📍 Original CRS for EM: EPSG:25833
📍 Original CRS for ES: EPSG:25833

📐 Original EM bounds (EPSG:25833): [13.18661414 52.4215107  13.53247313 52.57481712]
📐 Original ES bounds (EPSG:25833): [13.13425013 52.38922353 13.73461334 52.66074006]

🌍 Converted EM bounds (EPSG:4326): [1.05113743e+01 4.72809384e-04 1.05113774e+01 4.74192112e-04]
🌍 Converted ES bounds (EPSG:4326): [1.05113738e+01 4.72518174e-04 1.05113792e+01 4.74967083e-04]

⚠️ Invalid geometries - EM: 0, ES: 0

🔬 Sample EM geometry bounds (original): (13.347047220754405, 52.53890062771157, 13.365842948727332, 52.54616826557204)
🔬 Sample EM geometry bounds (converted): (10.511375691621726, 0.0004738681676160333, 10.511375860012608, 0.0004739337171898236)

🎯 Current calculated Berlin center: [52.52, 13.405]
🔍 Diagnostic complete - awaiting your approval for next step...


In [63]:
# 🛠️ Step 2: Fix CRS Labeling - The data is already WGS84!
print("🛠️ FIXING CRS: Correcting coordinate system labels...")
print("="*60)

# The diagnostic revealed the data is ALREADY in WGS84 (lat/lon) format
# Original bounds show Berlin coordinates: lat ~52.4-52.7, lon ~13.1-13.7
# But it's incorrectly labeled as EPSG:25833

# ✅ Correctly set CRS to WGS84 (no conversion needed!)
gdf_em_fixed = gdf_em.copy()
gdf_es_fixed = gdf_es.copy()

# Set the correct CRS - the data is already in WGS84 format
gdf_em_fixed.crs = "EPSG:4326"  # WGS84 - latitude/longitude
gdf_es_fixed.crs = "EPSG:4326"  # WGS84 - latitude/longitude

print(f"✅ EM CRS fixed: {gdf_em_fixed.crs}")
print(f"✅ ES CRS fixed: {gdf_es_fixed.crs}")

# 📍 Now calculate Berlin center correctly using the original coordinates
em_bounds_correct = gdf_em_fixed.total_bounds  # [minx, miny, maxx, maxy]
es_bounds_correct = gdf_es_fixed.total_bounds

print(f"\n📐 Corrected EM bounds (WGS84): {em_bounds_correct}")
print(f"📐 Corrected ES bounds (WGS84): {es_bounds_correct}")

# Combine bounds from both datasets
correct_minx = min(em_bounds_correct[0], es_bounds_correct[0])
correct_miny = min(em_bounds_correct[1], es_bounds_correct[1])
correct_maxx = max(em_bounds_correct[2], es_bounds_correct[2])
correct_maxy = max(em_bounds_correct[3], es_bounds_correct[3])

# Calculate center (lat, lon for Folium) 
berlin_center_fixed = [(correct_miny + correct_maxy) / 2, (correct_minx + correct_maxx) / 2]

print(f"\n🎯 Fixed Berlin center coordinates (lat, lon): {berlin_center_fixed}")
print(f"📍 Bounds: lon {correct_minx:.4f} to {correct_maxx:.4f}, lat {correct_miny:.4f} to {correct_maxy:.4f}")

# Validate coordinates are now correct for Berlin
if 52.3 <= berlin_center_fixed[0] <= 52.7 and 13.0 <= berlin_center_fixed[1] <= 13.8:
    print("✅ Coordinates look correct for Berlin!")
else:
    print("⚠️ Coordinates still look wrong...")

print("🚀 CRS fix complete - ready for mapping!")

🛠️ FIXING CRS: Correcting coordinate system labels...
✅ EM CRS fixed: EPSG:4326
✅ ES CRS fixed: EPSG:4326

📐 Corrected EM bounds (WGS84): [13.18661414 52.4215107  13.53247313 52.57481712]
📐 Corrected ES bounds (WGS84): [13.13425013 52.38922353 13.73461334 52.66074006]

🎯 Fixed Berlin center coordinates (lat, lon): [52.52498179802816, 13.434431733783263]
📍 Bounds: lon 13.1343 to 13.7346, lat 52.3892 to 52.6607
✅ Coordinates look correct for Berlin!
🚀 CRS fix complete - ready for mapping!


In [64]:
# 🗺️ Step 3: Create Interactive Folium Map
print("🗺️ CREATING MAP: Building interactive Folium visualization...")
print("="*60)

# 🎯 Create base map centered on Berlin with fixed coordinates
berlin_map = folium.Map(
    location=berlin_center_fixed,  # [lat, lon] = [52.525, 13.434]
    zoom_start=11,
    tiles='OpenStreetMap'
)

print(f"✅ Base map created, centered at: {berlin_center_fixed}")

# 🏠 Add EM Zones (Residential Protection) in BLUE
em_count = 0
for idx, row in gdf_em_fixed.iterrows():
    # Create popup with zone information
    popup_text = f"""
    <b>🏠 Residential Protection Zone</b><br>
    <b>Name:</b> {row['protection_zone_name']}<br>
    <b>District:</b> {row['district']}<br>
    <b>Area:</b> {row['area_ha']:.1f} hectares<br>
    <b>Effective:</b> {row['date_effective']}<br>
    <b>Zone Key:</b> {row['protection_zone_key']}
    """
    
    # Add polygon to map
    folium.GeoJson(
        row['geometry'],
        style_function=lambda x: {
            'fillColor': 'blue',
            'color': 'darkblue',
            'weight': 2,
            'fillOpacity': 0.6
        },
        popup=folium.Popup(popup_text, max_width=300)
    ).add_to(berlin_map)
    
    em_count += 1

print(f"✅ Added {em_count} EM zones (blue)")

# 🏙️ Add ES Zones (Urban Character Preservation) in GREEN  
es_count = 0
for idx, row in gdf_es_fixed.iterrows():
    # Create popup with zone information
    popup_text = f"""
    <b>🏙️ Urban Character Zone</b><br>
    <b>Name:</b> {row['protection_zone_name']}<br>
    <b>District:</b> {row['district']}<br>
    <b>Area:</b> {row['area_ha']:.1f} hectares<br>
    <b>Effective:</b> {row['date_effective']}<br>
    <b>Zone Key:</b> {row['protection_zone_key']}
    """
    
    # Add polygon to map
    folium.GeoJson(
        row['geometry'],
        style_function=lambda x: {
            'fillColor': 'green',
            'color': 'darkgreen', 
            'weight': 2,
            'fillOpacity': 0.6
        },
        popup=folium.Popup(popup_text, max_width=300)
    ).add_to(berlin_map)
    
    es_count += 1

print(f"✅ Added {es_count} ES zones (green)")

# 📍 Add a legend
legend_html = '''
<div style="position: fixed; 
     top: 10px; left: 50px; width: 200px; height: 90px; 
     background-color: white; border:2px solid grey; z-index:9999; 
     font-size:14px; padding: 10px">
<p><b>🗺️ Berlin Milieuschutz Zones</b></p>
<p><i class="fa fa-square" style="color:blue"></i> Residential Protection (EM)</p>
<p><i class="fa fa-square" style="color:green"></i> Urban Character (ES)</p>
</div>
'''
berlin_map.get_root().html.add_child(folium.Element(legend_html))

print(f"✅ Map complete with {em_count + es_count} total zones")
print("🖱️ Click on any zone for details!")

# Display the map
berlin_map

🗺️ CREATING MAP: Building interactive Folium visualization...
✅ Base map created, centered at: [52.52498179802816, 13.434431733783263]
✅ Added 81 EM zones (blue)
✅ Added 94 ES zones (green)
✅ Map complete with 175 total zones
🖱️ Click on any zone for details!
